src/OpenTelemetry.Extensions.Hosting/Implementation/OpenTelemetryMetricListener.cs

```diff
@@ -0,0 +1,120 @@
+// <copyright file="OpenTelemetryMetricListener.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+using Microsoft.Extensions.Diagnostics.Metrics;
+
+namespace OpenTelemetry.Metrics;
+
+internal sealed class OpenTelemetryMetricListener : IMetricsListener, IDisposable
+{
+    private readonly MeterProviderSdk meterProviderSdk;
+    private IObservableInstrumentsSource? observableInstrumentsSource;
+
+    public OpenTelemetryMetricListener(MeterProvider meterProvider)
+    {
+        var meterProviderSdk = meterProvider as MeterProviderSdk;
+
+        Debug.Assert(meterProviderSdk != null, "meterProvider was not MeterProviderSdk");
+
+        this.meterProviderSdk = meterProviderSdk!;
+
+        this.meterProviderSdk.OnCollectObservableInstruments += this.OnCollectObservableInstruments;
+    }
+
+    public string Name => "OpenTelemetry";
```

Comment:
What are the consequences of an app registering another listener by this same name? Is there any reason we should obscure a name a bit to reduce the chance of collision? `OpenTelemetrySdkMetricsListener`? Or does it not really matter?

Reply:
This name is important. It tells the engine how to bind settings for our listener. Similar to how in logging you can have different log levels for each `ILoggerProvider` you can have different metrics enabled for each `IMetricsListener`.

For example you can do things like this:

```json
"Metrics": {
   "EnabledMetrics": {
      "Default": false
   },
   "SomeMetricsSDK": {
      "EnabledMetrics": {
          "System.Net": true
      }
   },
   "OpenTelemetry": {
       "EnabledMetrics": {
           "Default": true
       }
   }
}
```

The name we use here is what users will need to put in their configuration.

Reply:
The actual registration we do is done via `TryAddEnumerable` so that there will only ever be one instance of `OpenTelemetryMetricsListener` regardless of the number of calls to `WithMetrics`. If someone else ships a listener named "OpenTelemetry" (shame on them 🤣) there will be two listeners bound to the same config but each will be managed correctly by the subscription manager.

To say that another way, the subscription manager operates on the [IEnumerable&lt;IMetricsListener&gt;](https://github.com/dotnet/runtime/blob/867b3e53094f2960f0bcc8b10ea2744bae538e0b/src/libraries/Microsoft.Extensions.Diagnostics/src/Metrics/MetricsSubscriptionManager.cs#L18C43-L18C63)s registered in the container. It doesn't make like a dictionary using the name(s). The name is only used to look up the [config/rules](https://github.com/dotnet/runtime/blob/867b3e53094f2960f0bcc8b10ea2744bae538e0b/src/libraries/Microsoft.Extensions.Diagnostics/src/Metrics/ListenerSubscription.cs#L133) that apply to the listener.